### PR TITLE
Sync Worklist To Saved Schedules

### PR DESCRIPTION
**Pull request by jaskirat-gill**
### Please provide a video demo below, or a screenshot and description of the change.

https://github.com/mlool/workday-calendar-extension/assets/60302907/3a1372e0-6400-4dba-9eed-7fe32c7bc0d6

https://github.com/mlool/workday-calendar-extension/assets/60302907/25b4c034-86f3-4727-842f-754310276450

### CODE REVIEW

File: src/content/ConfirmationModal/ConfirmationModal.css [hunk 1]
```diff
@@ -1,13 +1,13 @@
...
-  background-color: rgba(99, 99, 99, 0.6);
-  position: fixed;
-  top: 0;
-  height: 700px;
-  width: 300px;
-  justify-content: center;
-  align-items: center;
-  display: flex;
-  z-index: 100;
+    background-color: rgba(99, 99, 99, 0.6);
```
Comment (ThatNerdSquared) [inline src/content/ConfirmationModal/ConfirmationModal.css:2]: Some files like this one have weird indentation changes - a `yarn fmt-fix` should take care of them.

File: src/content/Form/Form.tsx [hunk 2]
```diff
@@ -1,17 +1,36 @@
...
+  const handleSyncClick = () => {
+    setIsSyncComponentVisible(true);
+  };
+
+  const closeSyncComponent = () => {
+    setIsSyncComponentVisible(false);
+  };
+
+  let sectionsForWorklist:ISectionData[] = []
+  for (const section of props.sections) {
```
Comment (ThatNerdSquared) [inline src/content/Form/Form.tsx:29]: I feel like this logic probably be extracted up to `App.tsx` as a `currentWorklistSections` handler - it's probably useful elsewhere in the codebase as well.

Also, ESLint errors

File: src/content/ModalLayer.tsx [hunk 3]
```diff
@@ -94,6 +103,65 @@ function ModalLayer(props: ModalLayerProps) {
...
+        )
+
+        return {
+          title: "Error Syncing With Saved Schedule",
+          body: errors,
+          closeButtonText: "Close",
+          actionButtonText: "Okay",
+          actionHandler: data.onConfirm,
+          cancelHandler: data.onCancel,
+          alignment: ModalAlignment.Center,
```
Comment (ThatNerdSquared) [inline src/content/ModalLayer.tsx:128]: These are nitpicks but:
- alignment is `Center` by default, so you can remove it here
- for the `actionButtonText`, I'd use "OK" instead of "Okay" (it just looks nicer, and modals usually use that lol)

File: src/content/SyncSavedSchedules/syncSavedSchedulesHelper.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    method: "POST",
+    body: urlencoded,
+    redirect: "follow" as RequestRedirect,
+  }
+
+  return fetch(
+    "https://wd10.myworkday.com/ubc/mass-select-action/2997$20406/13710$370.htmld",
+    requestOptions
+  )
+    .then((response) => response.json())
```
Comment (ThatNerdSquared) [inline src/content/SyncSavedSchedules/syncSavedSchedulesHelper.ts:33]: might be a good idea to use `await fetch` here, and refactor the following code to not use callbacks? they get kind of hard to follow once they're nested deeply.

**Review by ThatNerdSquared — COMMENTED** [review]
Just a few comments, looks pretty good. SyncSavedSchedules is a bit 🧐 but I know you're updating that so I didn't comment on it lol

File: src/content/ModalLayer.tsx [hunk 5]
```diff
@@ -45,6 +48,12 @@ interface ModalAction {
   additionalData?: unknown
 }
 
+interface SyncScheduleModalData {
+  data: [""]
```
[the remark this answers is not in the document]
Comment (jaskirat-gill) [reply src/content/ModalLayer.tsx:52]: oops yea did mean ``String[]`` Named it ``data`` because I was planning on using it to pass data for other sync modals but didnt end up using them, made it more descriptive now

File: src/content/ModalLayer.tsx (hunk 3 above)
Comment (jaskirat-gill) [reply src/content/ModalLayer.tsx:128]: valid points, fixed!

File: src/content/SyncSavedSchedules/syncSavedSchedulesHelper.ts (hunk 4 above)
Comment (jaskirat-gill) [reply src/content/SyncSavedSchedules/syncSavedSchedulesHelper.ts:33]: yea, the initial idea was to have just a working MVP that we can test and play with and nesting calls made debugging easier. I've went and refactored the way this function works and improved some variable naming

File: src/content/Form/Form.tsx (hunk 2 above)
Comment (jaskirat-gill) [reply src/content/Form/Form.tsx:29]: I like that idea. Went and extracted it all the way to ``util.ts`` as a utility function, ``filterSectionsByWorklist`` so that it can be called anywhere to filter without passing around extra props. For example, ``App.tsx`` now passes to ``Form.tsx`` ``sections={filterSectionsByWorklist(sections, currentWorklistNumber)}``

**Comment by jaskirat-gill** [comment]
@ThatNerdSquared thanks for the review. Implemented most of the feedback, there was actually not a lot that can be done for ``SyncSavedSchedules`` as it must be a React Component to use React hooks (``useContext`` for the Modals) but still cleaned it up best I could

File: src/content/Form/Form.tsx [hunk 6]
```diff
@@ -1,17 +1,29 @@
...
+  newSection: ISectionData | null;
+  sectionConflict: boolean;
+  handleAddNewSection: () => void;
+  handleCancel: () => void;
 }
 
 const Form = (props: IProps) => {
-  const dispatchModal = useContext(ModalDispatchContext)
+  const dispatchModal = useContext(ModalDispatchContext);
+  const [isSyncComponentVisible, setIsSyncComponentVisible] = useState(false);
```
Comment (ThatNerdSquared) [inline src/content/Form/Form.tsx:17]: this state isn't necessary anymore, the modal layer takes care of visibility once a modal is posted. same with the handlers below.

File: src/content/ModalLayer.tsx [hunk 7]
```diff
@@ -94,6 +103,62 @@ function ModalLayer(props: ModalLayerProps) {
...
+      case ModalPreset.SyncInstructions: {
+        const data: SyncScheduleModalData =
+          action.additionalData as SyncScheduleModalData
+
+        return {
+          title: "Sync Saved Schedules Instructions",
+          body: `Please note that you must be on the "View Saved Schedules" page. If you have multiple schedules, click the "add course sections" button on the one you which to add to, otherwise it will add to the first one. You must have all requirements (for example class requires lab and lecture) in your worklist`,
+          closeButtonText: "Close",
+          actionButtonText: "OK",
+          actionHandler: data.onConfirm,
```
Comment (ThatNerdSquared) [inline src/content/ModalLayer.tsx:141]: for this modal and the one below it, there's no actual action to be taken - you pass the same handlers for both the cancelHandler and the actionHandler. 

i'd suggest dropping all of the button-related properties - the modal layer will automatically add an "OK" button that closes the modal. See `Modal.HidePfpInfo` for an example

File: src/content/SyncSavedSchedules/SyncSavedSchedules.tsx [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+      }
+    } else {
+      alert(
+        `No saved schedule was detected. The most likely issue is that you are not on the "View My Saved Schedules" page. Please head to that page and try again.`
+      )
+      return
+    }
+  }
+
+  return <div></div>
```
Comment (ThatNerdSquared) [inline src/content/SyncSavedSchedules/SyncSavedSchedules.tsx:61]: why not move the button from the form here, so that it can show the first modal directly?

File: src/content/utils.ts [hunk 9]
```diff
@@ -283,3 +283,13 @@ const getTermFromSectionDetailsString = (
   if (includes2024) return Term.winterOne
   return Term.winterTwo
 }
+
+export const filterSectionsByWorklist = (sections: ISectionData[], worklist: number): ISectionData[] => {
+  const sectionsForWorklist:ISectionData[] = []
+  for (const section of sections) {
```
Comment (ThatNerdSquared) [inline src/content/utils.ts:289]: this is a nitpick, but you could've probably used `sections.filter` here. not a really necessary change though.
